cmd/chartsvc/handler_test.go

```diff
@@ -0,0 +1,666 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+)
+
+type bodyAPIListResponse struct {
+	Data apiListResponse `json:"data"`
+}
+
+type bodyAPIResponse struct {
+	Data apiResponse `json:"data"`
+}
+
+var chartsList []*models.Chart
+
+const testChartReadme = "# Quickstart\n\n```bash\nhelm install my-repo/my-chart\n```"
+const testChartValues = "image:\n  registry: docker.io\n  repository: my-repo/my-chart\n  tag: 0.1.0"
+
+func iconBytes() []byte {
+	var b bytes.Buffer
+	img := imaging.New(1, 1, color.White)
+	imaging.Encode(&b, img, imaging.PNG)
+	return b.Bytes()
+}
+
+func Test_chartAttributes(t *testing.T) {
+	tests := []struct {
+		name  string
+		chart models.Chart
+	}{
+		{"chart has no icon", models.Chart{
+			ID: "stable/wordpress",
+		}},
+		{"chart has a icon", models.Chart{
+			ID: "repo/mychart", RawIcon: iconBytes(),
+		}},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := chartAttributes(tt.chart)
+			assert.Equal(t, tt.chart.ID, c.ID)
+			assert.Equal(t, tt.chart.RawIcon, c.RawIcon)
+			if len(tt.chart.RawIcon) == 0 {
+				assert.Equal(t, len(c.Icon), 0, "icon url should be undefined")
+			} else {
+				assert.Equal(t, c.Icon, "/v1/assets/"+tt.chart.ID+"/logo-160x160-fit.png", "the icon url should be the same")
+			}
+		})
+	}
+}
+
+func Test_chartVersionAttributes(t *testing.T) {
+	tests := []struct {
+		name  string
+		chart models.Chart
+	}{
+		{"my-chart", models.Chart{
+			ID: "my-repo/my-chart", ChartVersions: []models.ChartVersion{{Version: "0.1.0"}},
+		}},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cv := chartVersionAttributes(tt.chart.ID, tt.chart.ChartVersions[0])
+			assert.Equal(t, cv.Version, tt.chart.ChartVersions[0].Version, "version string should be the same")
+			assert.Equal(t, cv.Readme, "/v1/assets/"+tt.chart.ID+"/versions/"+tt.chart.ChartVersions[0].Version+"/README.md", "README.md resource path should be the same")
+			assert.Equal(t, cv.Values, "/v1/assets/"+tt.chart.ID+"/versions/"+tt.chart.ChartVersions[0].Version+"/values.yaml", "values.yaml resource path should be the same")
+		})
+	}
+}
+
+func Test_newChartResponse(t *testing.T) {
+	tests := []struct {
+		name  string
+		chart models.Chart
+	}{
+		{"chart has only one version", models.Chart{
+			ID: "my-repo/my-chart", ChartVersions: []models.ChartVersion{{Version: "0.1.2"}}},
+		},
+		{"chart has many versions", models.Chart{
+			ID: "my-repo/my-chart", ChartVersions: []models.ChartVersion{{Version: "0.1.2"}, {Version: "0.1.0"}},
+		}},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cResponse := newChartResponse(&tt.chart)
+			assert.Equal(t, cResponse.Type, "chart", "response type is chart")
+			assert.Equal(t, cResponse.ID, tt.chart.ID, "chart ID should be the same")
+			assert.Equal(t, cResponse.Attributes.(models.Chart).ChartVersions, tt.chart.ChartVersions, "chart version in the response should be the same")
+		})
```

Comment:
can we check that the latestChartVersion relationship points to tt.chart.ChartVersions[0]?

Reply:
also the selfLink

Reply:
done!